File: python/src/main/python/Q017.py
from src.main.python.Solution import Solution

# Given a digit string, return all possible letter combinations that the number could represent.
#
# A mapping of digit to letters (just like on the telephone buttons) is given below.
#
# Input:    Digit string "23"
# Output:   ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].

# Note:
#   Although the above answer is in lexicographical order, your answer could be in any order you want.


class Q017(Solution):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        digit_to_letters = {'1': "",
                            '2': "abc",
                            '3': "def",
                            '4': "ghi",
                            '5': "jkl",
                            '6': "mno",
                            '7': "pqrs",
                            '8': "tuv",
                            '9': "wxyz",
                            '0': ""}

        if not digits:
            return []
        else:
            letters = digit_to_letters[digits[0]]
            subs = self.letterCombinations(digits[1:])
            ans = []
            for letter in letters:
                if subs:
                    for sub in subs:
                        ans.append(letter+sub)
                else:
                    ans.append(letter)
            return ans

File: java/src/main/java/solutions/Q017.java
package solutions;

import java.util.ArrayList;
import java.util.List;

/**
 * Given a digit string, return all possible letter combinations that the number could represent.
 *
 * Input:Digit string "23"
 * Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
 *
 * Note: Although the above answer is in lexicographical order, your answer could be in any order you want.
 */

public class Q017 extends Solution {

    public List<String> letterCombinations(String digits) {
        List<String> ans = new ArrayList<>();
        if (digits != null && !digits.isEmpty()) {
            letterCombinations(digits.toCharArray(), 0, new StringBuilder(), 0, ans);
        }
        return ans;
    }

    private void letterCombinations(char[] chars, int index, StringBuilder sb, int sbLen, List<String> ans) {
        if (index == chars.length) {
            ans.add(sb.toString());
        } else {
            char[] letters = mapDigit(chars[index]).toCharArray();
            for (char letter : letters) {
                sb.append(letter);
                letterCombinations(chars, index+1, sb, sbLen+1, ans);
                sb.setLength(sbLen);
            }
        }
    }

    private String mapDigit(char digit) {
        switch(digit) {
            case '2': return "abc";
            case '3': return "def";
            case '4': return "ghi";
            case '5': return "jkl";
            case '6': return "mno";
            case '7': return "pqrs";
            case '8': return "tuv";
            case '9': return "wxyz";
            default: return "";
        }
    }

}
